### ISSUE TITLE
compute: panicked at 'Invalid combination of type specializations: key types differ ...

### ISSUE THREAD

**Issue by mjibson** [MaterializeInc/materialize#22452]
Repro by starting locally with `--all-features`, execute

```
SELECT c.relname, NULL::pg_catalog.text FROM pg_catalog.pg_class c WHERE c.relkind IN ('v') AND (c.relname) LIKE '%' AND pg_catalog.pg_table_is_visible(c.oid) AND c.relnamespace <> (SELECT oid FROM pg_catalog.pg_namespace WHERE nspname = 'pg_catalog')
UNION ALL
SELECT NULL::pg_catalog.text, n.nspname FROM pg_catalog.pg_namespace n WHERE n.nspname LIKE '%' AND n.nspname NOT LIKE E'pg\\_%'
LIMIT 1000
```

Panic:
```
thread 'timely:work-0' panicked at 'Invalid combination of type specializations: key types differ! Row local vs. Bytes9 local, key_types: [ColumnType { scalar_type: Oid, nullable: false }]', /home/mjibson/materialize/src/compute/src/render/join/linear_join.rs:353:13
stack backtrace:
   0:     0x55c035e1a6f1 - std::backtrace_rs::backtrace::libunwind::trace::h66dc1c6acf794faa
                               at /rustc/d5c2e9c342b358556da91d61ed4133f6f50fc0c3/library/std/src/../../backtrace/src/backtrace/libunwind.rs:93:5
   1:     0x55c035e1a6f1 - std::backtrace_rs::backtrace::trace_unsynchronized::ha80d20099a67f790
                               at /rustc/d5c2e9c342b358556da91d61ed4133f6f50fc0c3/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
   2:     0x55c035e1a6f1 - std::sys_common::backtrace::_print_fmt::h7b959d43f35f16d4
                               at /rustc/d5c2e9c342b358556da91d61ed4133f6f50fc0c3/library/std/src/sys_common/backtrace.rs:65:5
   3:     0x55c035e1a6f1 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::hdaa196410d9ee0b9
                               at /rustc/d5c2e9c342b358556da91d61ed4133f6f50fc0c3/library/std/src/sys_common/backtrace.rs:44:22
   4:     0x55c035e4ffaf - core::fmt::rt::Argument::fmt::h0ddfbbe8be3f80d0
                               at /rustc/d5c2e9c342b358556da91d61ed4133f6f50fc0c3/library/core/src/fmt/rt.rs:138:9
   5:     0x55c035e4ffaf - core::fmt::write::h66b3c629f3d623e4
                               at /rustc/d5c2e9c342b358556da91d61ed4133f6f50fc0c3/library/core/src/fmt/mod.rs:1094:21
   6:     0x55c035e14837 - std::io::Write::write_fmt::hb6d80fba4115e0c2
                               at /rustc/d5c2e9c342b358556da91d61ed4133f6f50fc0c3/library/std/src/io/mod.rs:1714:15
   7:     0x55c035e1a505 - std::sys_common::backtrace::_print::h1a49cfb0cf3cce17
                               at /rustc/d5c2e9c342b358556da91d61ed4133f6f50fc0c3/library/std/src/sys_common/backtrace.rs:47:5
   8:     0x55c035e1a505 - std::sys_common::backtrace::print::hca95c2d0055e42a2
                               at /rustc/d5c2e9c342b358556da91d61ed4133f6f50fc0c3/library/std/src/sys_common/backtrace.rs:34:9
   9:     0x55c035e1c4f3 - std::panicking::default_hook::{{closure}}::hc03c01c56bca600c
                               at /rustc/d5c2e9c342b358556da91d61ed4133f6f50fc0c3/library/std/src/panicking.rs:269:22
  10:     0x55c035e1c284 - std::panicking::default_hook::hb2cb5315b6634f1c
                               at /rustc/d5c2e9c342b358556da91d61ed4133f6f50fc0c3/library/std/src/panicking.rs:288:9
  11:     0x55c034c95f39 - <alloc::boxed::Box<F,A> as core::ops::function::Fn<Args>>::call::h4e746e163c59bdad
                               at /rustc/d5c2e9c342b358556da91d61ed4133f6f50fc0c3/library/alloc/src/boxed.rs:2007:9
  12:     0x55c034bc7834 - mz_ore::panic::set_abort_on_panic::{{closure}}::h94c32382d9d98592
                               at /home/mjibson/materialize/src/ore/src/panic.rs:57:9
  13:     0x55c034c95f39 - <alloc::boxed::Box<F,A> as core::ops::function::Fn<Args>>::call::h4e746e163c59bdad
                               at /rustc/d5c2e9c342b358556da91d61ed4133f6f50fc0c3/library/alloc/src/boxed.rs:2007:9
  14:     0x55c032ca3b20 - mz_timely_util::panic::halt_on_timely_communication_panic::{{closure}}::hccbbda65d6fa7148
                               at /home/mjibson/materialize/src/timely-util/src/panic.rs:38:18
  15:     0x55c035e1cc5e - <alloc::boxed::Box<F,A> as core::ops::function::Fn<Args>>::call::hb3a915ffd78277c6
                               at /rustc/d5c2e9c342b358556da91d61ed4133f6f50fc0c3/library/alloc/src/boxed.rs:2007:9
  16:     0x55c035e1cc5e - std::panicking::rust_panic_with_hook::h75cd912a39a34e8a
                               at /rustc/d5c2e9c342b358556da91d61ed4133f6f50fc0c3/library/std/src/panicking.rs:709:13
  17:     0x55c035e1c9e7 - std::panicking::begin_panic_handler::{{closure}}::h1498b46f7849e167
                               at /rustc/d5c2e9c342b358556da91d61ed4133f6f50fc0c3/library/std/src/panicking.rs:597:13
  18:     0x55c035e1ab56 - std::sys_common::backtrace::__rust_end_short_backtrace::hd36a39b27b98086b
                               at /rustc/d5c2e9c342b358556da91d61ed4133f6f50fc0c3/library/std/src/sys_common/backtrace.rs:151:18
  19:     0x55c035e1c732 - rust_begin_unwind
                               at /rustc/d5c2e9c342b358556da91d61ed4133f6f50fc0c3/library/std/src/panicking.rs:593:5
  20:     0x55c035e4cac3 - core::panicking::panic_fmt::h98ef273141454c23
                               at /rustc/d5c2e9c342b358556da91d61ed4133f6f50fc0c3/library/core/src/panicking.rs:67:14
  21:     0x55c02d039ba8 - mz_compute::render::join::linear_join::dispatch_differential_join_inner_local_local::h64ed656a36642ec9
                               at /home/mjibson/materialize/src/compute/src/render/join/linear_join.rs:353:13
  22:     0x55c02d018c2a - mz_compute::render::join::linear_join::differential_join::ha664ecc1c8887b80
                               at /home/mjibson/materialize/src/compute/src/render/join/linear_join.rs:276:21
  23:     0x55c02d041464 - mz_compute::render::join::linear_join::<impl mz_compute::render::context::Context<G,T>>::render_join::{{closure}}::ha910409e2ae8523b
                               at /home/mjibson/materialize/src/compute/src/render/join/linear_join.rs:162:30
  24:     0x55c02ec0b468 - <timely::dataflow::scopes::child::Child<G,T> as timely::dataflow::scopes::Scope>::scoped::h6d2c55553078faf4
                               at /home/mjibson/.cargo/git/checkouts/timely-dataflow-70b80d81d6cabd62/53adc68/timely/src/dataflow/scopes/child.rs:129:13
  25:     0x55c02f05550d - timely::dataflow::scopes::Scope::region_named::h70a449bd33714c45
                               at /home/mjibson/.cargo/git/checkouts/timely-dataflow-70b80d81d6cabd62/53adc68/timely/src/dataflow/scopes/mod.rs:187:9
  26:     0x55c02ffb2eea - mz_compute::render::join::linear_join::<impl mz_compute::render::context::Context<G,T>>::render_join::h601a2a8c9596f844
                               at /home/mjibson/materialize/src/compute/src/render/join/linear_join.rs:101:9
  27:     0x55c03004084c - mz_compute::render::<impl mz_compute::render::context::Context<G>>::render_plan::h5d22cdbcaf983f19
                               at /home/mjibson/materialize/src/compute/src/render/mod.rs:931:25
  28:     0x55c03003fc83 - mz_compute::render::<impl mz_compute::render::context::Context<G>>::render_plan::h5d22cdbcaf983f19
                               at /home/mjibson/materialize/src/compute/src/render/mod.rs:888:29
  29:     0x55c03003fe0e - mz_compute::render::<impl mz_compute::render::context::Context<G>>::render_plan::h5d22cdbcaf983f19
                               at /home/mjibson/materialize/src/compute/src/render/mod.rs:892:28
  30:     0x55c03003fe0e - mz_compute::render::<impl mz_compute::render::context::Context<G>>::render_plan::h5d22cdbcaf983f19
                               at /home/mjibson/materialize/src/compute/src/render/mod.rs:892:28
  31:     0x55c03003fe0e - mz_compute::render::<impl mz_compute::render::context::Context<G>>::render_plan::h5d22cdbcaf983f19
                               at /home/mjibson/materialize/src/compute/src/render/mod.rs:892:28
  32:     0x55c03003fe0e - mz_compute::render::<impl mz_compute::render::context::Context<G>>::render_plan::h5d22cdbcaf983f19
                               at /home/mjibson/materialize/src/compute/src/render/mod.rs:892:28
  33:     0x55c03003fe0e - mz_compute::render::<impl mz_compute::render::context::Context<G>>::render_plan::h5d22cdbcaf983f19
                               at /home/mjibson/materialize/src/compute/src/render/mod.rs:892:28
  34:     0x55c03003fe0e - mz_compute::render::<impl mz_compute::render::context::Context<G>>::render_plan::h5d22cdbcaf983f19
                               at /home/mjibson/materialize/src/compute/src/render/mod.rs:892:28
  35:     0x55c0300452ae - mz_compute::render::<impl mz_compute::render::context::Context<G>>::build_object::hda0fafb239dc9101
                               at /home/mjibson/materialize/src/compute/src/render/mod.rs:490:22
  36:     0x55c02d96a817 - mz_compute::render::build_compute_dataflow::{{closure}}::{{closure}}::h37de2858aea6e714
                               at /home/mjibson/materialize/src/compute/src/render/mod.rs:385:21
  37:     0x55c02ec115c0 - <timely::dataflow::scopes::child::Child<G,T> as timely::dataflow::scopes::Scope>::scoped::he433eb4640eff3b9
                               at /home/mjibson/.cargo/git/checkouts/timely-dataflow-70b80d81d6cabd62/53adc68/timely/src/dataflow/scopes/child.rs:129:13
  38:     0x55c02f0555c6 - timely::dataflow::scopes::Scope::region_named::hc78cd51dc339a1d4
                               at /home/mjibson/.cargo/git/checkouts/timely-dataflow-70b80d81d6cabd62/53adc68/timely/src/dataflow/scopes/mod.rs:187:9
  39:     0x55c02d969829 - mz_compute::render::build_compute_dataflow::{{closure}}::he5d2f374ade56f98
                               at /home/mjibson/materialize/src/compute/src/render/mod.rs:352:13
  40:     0x55c031210523 - timely::worker::Worker<A>::dataflow_core::hba35e8b7e508b61b
                               at /home/mjibson/.cargo/git/checkouts/timely-dataflow-70b80d81d6cabd62/53adc68/timely/src/worker.rs:639:13
  41:     0x55c02d96917d - mz_compute::render::build_compute_dataflow::h4382ecde157b22fe
                               at /home/mjibson/materialize/src/compute/src/render/mod.rs:198:5
  42:     0x55c02d93e29d - mz_compute::compute_state::ActiveComputeState<A>::handle_create_dataflow::h533451e45ebe0246
                               at /home/mjibson/materialize/src/compute/src/compute_state.rs:278:9
  43:     0x55c02d93dd2d - mz_compute::compute_state::ActiveComputeState<A>::handle_compute_command::hd66c4f8e03fae6db
                               at /home/mjibson/materialize/src/compute/src/compute_state.rs:181:41
  44:     0x55c02d9b8bfe - mz_compute::server::Worker<A>::handle_command::hba17be699328009d
                               at /home/mjibson/materialize/src/compute/src/server.rs:424:9
  45:     0x55c02d9b7bc9 - mz_compute::server::Worker<A>::run_client::h48cf13cfac29b9e0
                               at /home/mjibson/materialize/src/compute/src/server.rs:402:17
  46:     0x55c02d9b8f95 - mz_compute::server::Worker<A>::setup_channel_and_run_client::h2fa2f454e6ca91f0
                               at /home/mjibson/materialize/src/compute/src/server.rs:350:9
  47:     0x55c02d9b9d33 - mz_compute::server::Worker<A>::run::h6f9bfb70fec8707c
                               at /home/mjibson/materialize/src/compute/src/server.rs:214:21
...
```

**Comment by mjibson** [comment]
(I am working on getting the exact query psql is running, will update this when I have it.)

**Comment by antiguru** [comment]
@vmarcos Can you figure out why the feature is on in @mjibson's repro?

**Comment by mjibson** [comment]
I'm running with `--all-features`!

**Comment by mjibson** [comment]
Crashing query is:

```
SELECT c.relname, NULL::pg_catalog.text FROM pg_catalog.pg_class c WHERE c.relkind IN ('v') AND (c.relname) LIKE '%' AND pg_catalog.pg_table_is_visible(c.oid) AND c.relnamespace <> (SELECT oid FROM pg_catalog.pg_namespace WHERE nspname = 'pg_catalog')
UNION ALL
SELECT NULL::pg_catalog.text, n.nspname FROM pg_catalog.pg_namespace n WHERE n.nspname LIKE '%' AND n.nspname NOT LIKE E'pg\\_%'
LIMIT 1000
```

**Comment by antiguru** [comment]
OK, we have to disable the arrangement specialization feature for `--all-features` because it's known to not work for multiple joins.

**Comment by antiguru** [comment]
Candidate fix in #22453

**Comment by moulimukherjee** [comment]
Not a release blocker, removed the tag.
